src/samples/http/server-sse.c

```diff
@@ -0,0 +1,210 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2015 Intel Corporation. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @file
+ * @brief HTTP Server sse
+ *
+ * Sample server that return a response and keep it alive.
+ * It implements server sent events. The content of this
+ * is read from stdin. To see the usage help, -h or --help.
+ */
+
+
+#include <getopt.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "soletta.h"
+#include "sol-http.h"
+#include "sol-http-server.h"
+#include "sol-util.h"
+#include "sol-util-file.h"
+
+static struct sol_http_server *server;
+static struct sol_fd *stdin_watch;
+static struct sol_buffer value = SOL_BUFFER_INIT_EMPTY;
+static struct sol_ptr_vector responses = SOL_PTR_VECTOR_INIT;
+
+static bool
+on_stdin(void *data, int fd, uint32_t flags)
+{
+    uint16_t i;
+    struct sol_http_response_sse *sse;
+
+    if (flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP)) {
+        fprintf(stderr, "ERROR: Something wrong happened with file descriptor: %d\n", fd);
+        goto err;
+    }
+
+    if (flags & SOL_FD_FLAGS_IN) {
+        int err;
+
+        value.used = 0;
+        err = sol_buffer_insert_bytes(&value, 0,
+            (uint8_t *)"data: ", sizeof("data: ") - 1);
+        if (err < 0) {
+            fprintf(stderr, "ERROR: failed to append the data prefix: %s\n",
+                sol_util_strerrora(-err));
+            goto err;
+        }
+        /* this will loop trying to read as much data as possible to buffer. */
+        err = sol_util_load_file_fd_buffer(fd, &value);
+        if (err < 0) {
+            fprintf(stderr, "ERROR: failed to read from stdin: %s\n",
+                sol_util_strerrora(-err));
+            goto err;
+        }
+
+        if (value.used == 0) {
+            /* no data usually means ^D on the terminal, quit the application */
+            printf("no data on stdin, quitting.\n");
+            sol_quit();
```

Comment:
this won't work, if the server did not send all the data, the mainloop will exit and client won't receive it all.

You should call `sol_quit()` from your callback to response "on_close". And here is a problem with your implementation: when user explicitly called `sse_response_del()`, you do not call his callback, then there is no way to know when the data was fully sent to client. Actually, since you mark `delete_me`, it will exit even if not all data was sent. This whole thing must be reworked so we can abort (close immediately) or close after flush/done.